src/input/scan.rs

```diff
@@ -0,0 +1,213 @@
+use log::debug;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+pub const INITIAL_DEVS_AVAILABLE_FOR: Duration = Duration::from_millis(1000);
+
+lazy_static! {
+    /// A singleton of the EvDevsScan object
+    pub static ref SCANNED: EvDevs = EvDevs::new();
+}
+
+/// This struct contains the results of initially scaning all evdev devices,
+/// which allows for device model independancy.
+/// Some of its data is used by other constants.
+///
+/// EvDevsScan has some internal mutability to allow resuing the opened devices
+/// for some time to increase performance.
+/// TODO: Call this `EvDevsScanOutcome` or EvScanOutcome instead ??
+pub struct EvDevs {
+    pub wacom_path: PathBuf,
+    pub multitouch_path: PathBuf,
+    pub gpio_path: PathBuf,
+
+    pub wacom_width: u16,
+    pub wacom_height: u16,
+    pub mt_width: u16,
+    pub mt_height: u16,
+
+    /// The resolution of the wacom no rotation applied
+    pub wacom_orig_size: Vector2<u16>,
+    pub multitouch_orig_size: Vector2<u16>,
+
+    // Those will be preserved in case they are needed fairly fast
+    // to prevent any additional delay of re-opening the fds.
+    // They will get removed fairly quickly though.
+    wacom_initial_dev: Arc<Mutex<Option<evdev::Device>>>,
+    multitouch_initial_dev: Arc<Mutex<Option<evdev::Device>>>,
+    gpio_initial_dev: Arc<Mutex<Option<evdev::Device>>>,
+}
+
+impl EvDevs {
+    /// Scan all the evdev devices, figure out which is which
+    /// and get some additional data for lazy constants.
+    fn new() -> Self {
+        // All of these have to be found
+        let mut wacom_path: Option<PathBuf> = None;
+        let mut wacom_dev: Option<evdev::Device> = None;
+        let mut multitouch_path: Option<PathBuf> = None;
+        let mut multitouch_dev: Option<evdev::Device> = None;
+        let mut gpio_path: Option<PathBuf> = None;
+        let mut gpio_dev: Option<evdev::Device> = None;
+
+        // Get all /dev/input/event* file paths
+        let mut event_file_paths: Vec<PathBuf> = Vec::new();
+        let input_dir = Path::new("/dev/input");
+        for entry in input_dir
+            .read_dir()
+            .unwrap_or_else(|_| panic!("Failed to list {:?}", input_dir))
+        {
+            let entry = entry.unwrap();
+            let file_name = entry.file_name().as_os_str().to_str().unwrap().to_owned();
+            if !file_name.starts_with("event") {
+                continue;
+            }
+
+            let evdev_path = input_dir.join(&file_name);
+            event_file_paths.push(evdev_path);
+        }
+
+        // Open and check capabilities of each event device
+        for evdev_path in event_file_paths {
+            let dev = evdev::Device::open(&evdev_path)
+                .unwrap_or_else(|_| panic!("Failed to scan {:?}", &evdev_path));
+            if dev.events_supported().contains(evdev::KEY) {
+                if dev.keys_supported().contains(evdev::BTN_STYLUS as usize)
```

Comment:
Could a first check of `wacom_dev.is_none() && ...` help with the timings you're experiencing?

Same with the other 2  input devices.

Reply:
Not really. We still have 3 /dev/input/eventX files and each is opened once here, figured out what it is, and the pair of `_path` and `_dev` get assigned accordingly.

The problem is, that the new hardware mixed up the oder. See: https://github.com/Eeems/oxide/issues/48#issuecomment-690830572

So that loop is opening every device and checking the capabilities to discover what device is which InputDevice.